src/main/java/io/spokestack/spokestack/nlu/NLUService.java

```diff
@@ -0,0 +1,52 @@
+package io.spokestack.spokestack.nlu;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A base class for components that provide intent classification and slot
+ * recognition, either on-device or via a network request.
+ *
+ * <p>
+ * Intent classification is designed to be an asynchronous process; in order
+ * to receive results as they are available, add a listener to the NLU
+ * service.
+ * </p>
+ */
+public abstract class NLUService {
```

Comment:
I think it would be better to decouple the default implementation from the nlu interface by creating a java interface for it, along with whatever base implementation you want to provide
this way, clients aren't exposed to implementation details like the public `dispatch` method below
I would probably go further and just provide the interface without the anemic base class, since there won't be many implementations

Reply:
I actually want this to be just an interface, but I ended up with this compromise because I don't have an object like `SpeechContext` here that threads through the components and handles the event dispatching. I didn't want NLU service providers to have to reimplement event dispatching every time—and I don't know whether this is really a "base class" since you can't instantiate it; you have to implement `classify`.

That said, there's probably another design option I'm missing here; open to feedback.

Reply:
every abstract class is a "base class"
implementing event subscription/dispatching is trivial, but if you insist on making this a class, at least make `dispatch` protected so it's not in the public interface